File: Algorithms/Greedy_algorithms/fractional_knapsack.c
// Given a set of items with their weights and values, fill a knapsack of limited capacity with items
// to maximize the total value, considering fractions of items

#include <stdio.h>

void fractionalKnapsack(int weights[], int values[], int n, int capacity) {
    double valuePerWeight[n];
    for (int i = 0; i < n; i++) {
        valuePerWeight[i] = (double) values[i] / weights[i];
    }

    for (int i = 0; i < n - 1; i++) {
        int maxIndex = i;
        for (int j = i + 1; j < n; j++) {
            if (valuePerWeight[j] > valuePerWeight[maxIndex]) {
                maxIndex = j;
            }
        }
        double temp = valuePerWeight[i];
        valuePerWeight[i] = valuePerWeight[maxIndex];
        valuePerWeight[maxIndex] = temp;

        int temp2 = weights[i];
        weights[i] = weights[maxIndex];
        weights[maxIndex] = temp2;

        temp2 = values[i];
        values[i] = values[maxIndex];
        values[maxIndex] = temp2;
    }

    int currentWeight = 0;
    double finalValue = 0.0;

    for (int i = 0; i < n; i++) {
        if (currentWeight + weights[i] <= capacity) {
            currentWeight += weights[i];
            finalValue += values[i];
        } else {
            int remainingCapacity = capacity - currentWeight;
            finalValue += valuePerWeight[i] * remainingCapacity;
            break;
        }
    }

    printf("Maximum value in knapsack: %.2lf\n", finalValue);
}

int main() {
    int weights[] = {10, 20, 30};
    int values[] = {60, 100, 120};
    int capacity = 50;
    int n = sizeof(weights) / sizeof(weights[0]);
    fractionalKnapsack(weights, values, n, capacity);
    printf("\n");
    return 0;
}

File: Algorithms/Greedy_algorithms/fractional_knapsack.py
# Given a set of items with their weights and values, fill a knapsack of limited capacity with items
# to maximize the total value, considering fractions of items

def fractional_knapsack(weights, values, capacity):
    n = len(weights)
    value_per_weight = [v / w for v, w in zip(values, weights)]

    sorted_indices = sorted(range(n), key=lambda i: -value_per_weight[i])

    current_weight = 0
    final_value = 0

    for i in sorted_indices:
        if current_weight + weights[i] <= capacity:
            current_weight += weights[i]
            final_value += values[i]
        else:
            remaining_capacity = capacity - current_weight
            final_value += value_per_weight[i] * remaining_capacity
            break

    print("Maximum value in knapsack:", final_value)

weights = [10, 20, 30]
values = [60, 100, 120]
capacity = 50
fractional_knapsack(weights, values, capacity)
